### DISCUSSION TITLE
Ideas Collection

### DISCUSSION THREAD

**Discussion by antfu** [nuxt/devtools#29] [Ideas]
If you have any ideas/requests for features builtin into Nuxt DevTools, feel free to brainstorm your ideas below! 🙌

**Comment by OmarMcAdam** [comment]
**Component Renderer**

- Visually see a component rendered within the Dev Tools
- Have Inputs for each of the props
   - If props are typed, set input controls to match (`bool` = toggle, `number` = number field, `string` = text field .... etc)

Usecase:
This will help developers by creating a visual library of all the components and be able to rapidly test props to reach desired effect.
More suitable for UI library style components rather than data heavy items.

**Reply by OlaAlsaker** [reply]
What you are describing is [Storybook](https://storybook.js.org) 😄

**Reply by OmarMcAdam** [reply]
> What you are describing is [Storybook](https://storybook.js.org) 😄

So it is.... I've been avoiding their ads like the plague
Might need to actually give them a go then

**Reply by OlaAlsaker** [reply]
Yes, I recommend checking it out! However, `nuxt/storybook` is still not compatible with Nuxt 3. You are welcome to follow the discussion here: https://github.com/nuxt-community/storybook/issues/330

**Reply by antfu** [reply]
We are talking with @Akryum to have [Histoire](https://histoire.dev/) module contributing a tab for that!

**Reply by OlaAlsaker** [reply]
Cool, never heard about it before – maybe I should check it out 😄

**Reply by grupoproativa** [reply]
pessoal, preciso de um aplicativo que conte numero de comentarios no instagram, tipo, numero de @ mais marcado, comentarios repetidos e etc: alguem pode me ajudar? eu já vi um semelhante, se alguem puder me ajudar, entra em contato 81 9 9173-9117

**Comment by BracketJohn** [comment]
> **Note**: This feature is supported in the latest version ✅

**API Playground**

A swagger-style API playground would be great. This is something that the Python framework FastAPI has implemented and the Python community has been super drawn by it. I think it would be a great addition to Nuxt for:
- PMs to test features without UI already being present,
- Devs to iterate on endpoints
- PMs/... showing off API-functionality to non-technical stakeholders

Here's what a swagger API playground would look like (taken from https://fastapi.tiangolo.com/features/)
<img width="736" alt="image" src="https://user-images.githubusercontent.com/25911230/217891641-40acfd60-612a-421b-9453-051b29e20f63.png">

Here's a prior issue on this from nitro: https://github.com/unjs/nitro/issues/4

Thanks for these devtools, they are already great! 🦃

**Reply by jd-solanki** [reply]
Love to see FastAPI here ❤️

**Reply by antfu** [reply]
Thanks! We already have it in plan 🙌

**Reply by atinux** [reply]
Glad to see I was not alone thinking of this :D

**Reply by BracketJohn** [reply]
:D It has been in the back of my mind since end of 2021 when we started to switch from nuxt 2 + python backends to fullstack nuxt 3 :D

**Comment by ThimoDEV** [comment]
**Profiler & Rerender highlighter**

Just like the react devtools a functionality to record actions and see how much of a performance impact it has. See when components get mounted/unmounted etc so you know that component can be more optimized. I don't know if this is possible..

**Reply by antfu** [reply]
We are trying to add performance metrics here and there bit by bit. As such metrics can be hard to collect at runtime.

Current we have

- Time to render/navigate the current page, via the floating panel
- Time to build, bundle size, via the "Analyze Build" tab
- Hooks time, via "Hooks" tab

For mounted/unmounted time, it feels more low level at runtime, maybe better belongs to Vue DevTools.

**Comment by harlan-zw** [comment]
Built-in logger! Context: https://github.com/nuxt/nuxt/issues/18701

@danielroe did you by any chance start working on this? Thought I may have seen a related tweet :thinking:

**Reply by atinux** [reply]
I guess we could already hook into `consola` (cc @pi0 )

**Reply by danielroe** [reply]
I have indeed! Will open a pr shortly with initial implementation and update issue with my recommendations for next steps.

**Comment by francisashley** [comment]
> **Note**: This feature is supported in the latest version, with #282 and #207  ✅

Firstly .. this tool is amazing!

In terms of DX, it would be great if:

- we can drag the overlay around the page
- resize the overlay
- resize panels inside the overlay
- reposition the toggle
- hide the toggle
- open components/page/plugins/file paths etc in vscode via`option/alt+click` https://stackoverflow.com/a/53786004/18133925

**Reply by antfu** [reply]
Thanks! I think most of them are addressed with https://github.com/nuxt/devtools/pull/282 and https://github.com/nuxt/devtools/pull/207.

**Reply by davidwebca** [reply]
Hi @antfu! Is there an internal API / function that allows us to tap into that component path? I'm trying to create a module that looks into a "queries" folder and fetches automatically any ".gql" graphql file that corresponds to the component's path/name. I have a very simple proof of concept as a composable ([see here](https://gist.github.com/davidwebca/d94c1d0762c2f8e9149fbab690a96870)) that does somewhat work in dev, many tests to do stil of coursel, but the main thing I hate is how I resolve the path because it currently doesn't work while working in the module playground because it's under the "/playground" directory. I could simply do a "replace" again but all of these are dirty and prone to error in end-user land. I'm searching the codebase and docs right now to maybe get the proper component's path relative to the root directory, exactly what's currently found in the "data-v-inspector" attribute. 🤔 

I might not be looking and asking at the right places, any tip would be appreciated. 🤘

**Comment by bbcvc** [comment]
> **Note**: This feature is supported in the latest version, with #282 ✅

### About dock side
Add a feature for `Undock into separate window`.
like Chrome devtools
![image](https://user-images.githubusercontent.com/52622877/218656739-4faa529e-17aa-4f3c-9fed-93402e5cec71.png)

**Reply by devzom** [reply]
this would be great!

**Reply by antfu** [reply]
Supported with #282, thanks!

**Reply by AdrianoCahete** [reply]
Can we dock to the sides? If yes, how?

If not, there's a request for that already?

**Comment by zyyv** [comment]
> **Note**: This feature is supported in the latest version, with #120 ✅

Thanks to this powerful tool, our development experience has been further improved!

How about add static resource analysis? Usage like `Componets`.

It can achieve the following these: 
- We can locate and know that the resource is used in that component or css file.
- Extends the file tree structure and preview it.
- Tile all resources and highlight used resources.
- etc.

The above suggestions are all referenced by images.

**Reply by zyyv** [reply]
Implemented in #120.

**Comment by cpreston321** [comment]
> **Note**: This feature is supported in the latest version, with #194 and #247 ✅

As a user, 

I would like to access [nuxt docs](https://nuxt.com/docs/getting-started/introduction) within the dev tools without context switching between tabs to search through docs or maybe a more elegant way to be contained within the dev tools view.

I appreciate this powerful tool and all the hard work! 💚

**Reply by antfu** [reply]
For sure! It's on our roadmap already - we just need to figure out a good way to make it interactive.

**Reply by ippatev** [reply]
What do you think about a similar solution like this: https://supabase.com/blog/openai-embeddings-postgres-vector

**Reply by ddahan** [reply]
Based on the same idea, I would love having access to [Nuxt Labs UI doc](https://ui.nuxtlabs.com/getting-started) too.
This level of deep integration could be a reason to favor "official" libraries rather than external ones.

**Reply by Jak2k** [reply]
Maybe, a way to add *any* website woulde be a good idea. I'm currently using custom tabs in the nuxt-config, but the vue-docs don't allow to be embeded.

**Reply by antfu** [reply]
Implemented with https://github.com/nuxt/devtools/pull/194 and https://github.com/nuxt/devtools/pull/247

**Comment by szulcus** [comment]
In Windows 11, when I try to open file in VS Code:
```
'C:\Users\Jakub' is not recognized as an internal or external command,
operable program or batch file.
```

**Reply by antfu** [reply]
I guess it might not be related to Nuxt DevTools. Please open an issue with reproductions for bug reports.

**Comment by paulmelero** [comment]
Runtime Config inspection would be awesome, but being able to edit the values in runtime would be even more awesome.

I showcased this in a demo recently: https://github.com/gangsthub/nuxt-module-example-custom-devtools/

But the problem is that I couldn't make it work without changing [how you access the values in the code](https://www.linkedin.com/in/paulmelero/).

**Comment by paulmelero** [comment]
What about having also a requests interceptor? (using [msw.js](https://mswjs.io/) or similar)

Or maybe it's an opportunity for a separate module that integrates with the Nuxt DevTools.

**Comment by ManUtopiK** [comment]
Record the browser activity and generate appropriate test files.
Use https://github.com/Pythagora-io/pythagora for server side activities.

**Comment by cpreston321** [comment]
# Examine Project 

As a user, 

I would like dev tools to examine the project to try to look for best practices and ways to improve my experience. This one might be a little bit more tricky, I think this would useful. Showing ways to better improve the app in general and in performance. The only problem is trying to figure out what's right. 

## Example

- E.g: If you use SSG over node server you could possibly see a performance increase on average % -%.
- E.g: Instead of using `thisFunction` have you tried possibly making into a composable? You have duplicate code here, here and here. The downside to this is....
- E.g You don't seem to be using  `server/`, it might be optimal to delete this directory until it's used.
- E.g The Nuxt Team prefers using `....` over `....`
- E.g It seems that you using the options api with only one method and data, have you tried using the Composition API to even simply it more ?

## Options

The standard is set based on what version of nuxt so then make it easier when people upgrade they can see where functions are deprecated or there is better alternatives.

- **Full Examine** 
  - Takes quite a bit longer but can be a button that pressed to then scan the whole repo itself with some progress bar.
- **Quick Examine** (Run on start of dev)
  - Maybe it just scan components, pages, endpoints that loaded initially AKA (whatever page they are currently on).


## Conclusion

The goal is to show how do best practices, optimize code/quality and performance. This could help people greatly especially for beginners and just old projects migrating to the newer standards of nuxt.

**Comment by SuleymanDeveloper** [comment]
## It would be nice if here can filter with folders.
<img width="927" alt="Screenshot 2023-02-18 at 21 32 12" src="https://user-images.githubusercontent.com/82510192/219877307-99cd54ba-ab89-4c94-8cc2-96654f1a3d8d.png">

**Comment by tesla3327** [comment]
> **Note**: This feature is supported in the latest version, with #100 ✅

**Inspect/modify storage**

The hardest part of working with the storage layer of Nitro is that it's hard to know what's actually in there.

A visual editor of some sort would be fantastic.

**Reply by antfu** [reply]
@Atinux implemented it in https://github.com/nuxt/devtools/pull/100, you need to enable `experimental features` to see it. Let us know what do you see for improvements.

**Comment by mgd216** [comment]
Amazing tool thank you !

Pinia:  Is there someone already working on a Pinia module so that you can see the store state?

**Reply by markedro** [reply]
I was wondering the same. Have been looking around to see if I was just missing it but didn't find it. Also tried to get it working in the vue devtools chrome extension v6.5 but no luck there either so far. Viewing the current state (as well as modify it) is the #1 thing I'd imagine getting used for the folks using Pinia.

**Reply by Zendrex** [reply]
Bumping this, also looking forward to having Pinia support in the new devtools. Thought I was missing it, but alas it just isnt integrated.

**Reply by toeyeen** [reply]
You have access to the Pinia module via vite-plugin-vue-devtools for the Vue, but the dev tool for Nuxt doesn't show the Pinia Module.

**Reply by antfu** [reply]
Could you share more details about what do you expect about the Pinia UI to provide. Currently, Pinia already provides a state viewer and editor in Vue DevTools. 

@arashsheyda is working on a Pinia module for that but we are wondering what else we could provide other than the features already in Vue DevTools

**Reply by therealokoro** [reply]
Any update on this please?

**Comment by Oldwo1f** [comment]
### Composables Tab

There is no tab for composable.
It would be nice to 
1. see the properties in real time (possibly havethe option to edit these values)
2. see the methods to get clue of their arguments (possibly have a button to trigger the method)

**Reply by antfu** [reply]
Currently, we have `imports` tab listing all the composables. Getting arguments from runtime function calls can be very challenging. Feel free to send draft PR if you have any solution ideas so we can work together. Thanks!

**Comment by ManUtopiK** [comment]
### Inspector

An option to toggle inspector mode and not pollute the DOM :

![image](https://user-images.githubusercontent.com/188172/227089320-1e7ba0e9-340b-4105-9eb3-06c7ee2c3480.png)

**Reply by antfu** [reply]
You can disable it by

```ts
export default defineNuxtConfig({
  devtools: {
    componentInspector: false
  }
})
```

**Reply by Jak2k** [reply]
Toogling it in realtime would be nice, but this is also ok. 😆

**Comment by TechWatching** [comment]
Are there plans to integrate some features of the Vue dev tools like the timeline or the live view  of the props, refs, computed values? Both tools are great but it would feel weird to have to open Vue dev tools for some features and nuxt dev tools for other.

**Reply by antfu** [reply]
Nuxt DevTools is designed to work alongside Vue DevTools (they cover different scopes). In the short term, we will focus on what we could provide for Nuxt-specific integrations, instead of mimic the features already had in Vue DevTools. In the long term, we will see how we could make them work together more organically.

**Comment by Suniron** [comment]
Make it work in Nuxt Bridge project ! 🙏

**Reply by antfu** [reply]
Thanks for bringing this up. As there is a lot to explore in DevTools, I think currently we will put our focus on the DevTools itself. Porting to Bridge might not be in our roadmap very soon.

**Comment by arashsheyda** [comment]
I was thinking is it a good idea to add contextmenu to assets tab? we could have actions like refresh or rename and delete and ...
and we can useMagicKeys for some shortcuts?
[example video](https://res.cloudinary.com/arashsheyda/video/upload/v1680514967/nuxt/devtool-contextmenu_eufqn7.mp4)

**Reply by antfu** [reply]
I would normally avoid touching contextmenu if possible. For actions we could put them in the details etc.

**Reply by arashsheyda** [reply]
hmmm okay then, thank you

**Comment by arashsheyda** [comment]
> **Note**: This feature is supported in the latest version, with #222 ✅

what if we add `marketplace` to devtools? to manage our modules and themes(community ones and official nuxt-themes)?
Is it a good idea to work on?

![demo](https://user-images.githubusercontent.com/38922203/230653747-95dc1e52-0a3e-4e6e-9fb3-95f3d475a476.png)

**Reply by arashsheyda** [reply]
here is a experimental [demo video](https://res.cloudinary.com/arashsheyda/video/upload/v1680967480/nuxt/devtool-marketpalce-exp-demo_o2uwvn.mp4)

**Reply by antfu** [reply]
For modules, we already have working on integrating it into the modules tab:
https://twitter.com/Atinux/status/1643556695786790912

For themes and plugins, can you explain a bit more about how do you think we could have them?

**Reply by arashsheyda** [reply]
@antfu Oh that's cool thanks.
For themes i think we can do something like modules, where there is official themes like docus and comunity ones and we can just add it and extend it to our project and maybe have a place for submitting new themes directly from devtool?

for plugins I'm not sure maybe ready written plugin files(now that i'm thinking it's not that necessary lol)

**Reply by Jak2k** [reply]
Writing plugins is a pain. 😆

**Comment by lif3ng** [comment]
> Note: This feature is supported in the latest version, with #282 ✅

An optional view mode: when show/dock/resize devtools， change `<body>` or `#__nuxt` style，like a playground page

For example:
![image](https://user-images.githubusercontent.com/8555501/232233660-77c922e6-4dae-449a-b893-1974d7fd6ab8.png)

**Reply by antfu** [reply]
I will see if we can find a way to make Nuxt DevTools work inside DevTools as a Chrome extension. Which I think would be better than injecting scripts that might break the app.

**Reply by antfu** [reply]
This should be solved with #282. Thanks!

**Comment by Hebilicious** [comment]
Thanks for making something so incredible !

Something that I would personally love to see is a Vite tab (similar to the Inspect tab).
In Nuxt it can be hard to know what's going on with the vite config, because it can be modified from multiple places (framework, nuxt.config.ts, modules ...)
A tab that display a clear list of which vite settings/plugins are enabled, maybe with a way to see their current config, the order they run in and a link to their source file (plugin A is coming from module B) would be really great.
I'm not sure how complex of a task that is though...

Just went through this process while creating a module that adds markdown pages support for nuxt, and figuring out what was going on with the vite config was really tedious. A lot of console logging inside node modules and node debugging was necessary.

**Reply by antfu** [reply]
I am not very should it could be helpful, or it can be very tricky to implement in a good way. I think displaying the **resolve config** (it will be hard to read tho) statically should be doable. But knowing the source would require quite a complex AST transform.

I feel this might go out of Nuxt DevTools' scope. If we even have a good presentation UI for this, I think we could have it in vite-plugin-inspect.

Thanks for bringing this up.

**Comment by Jak2k** [comment]
Displaying a QR-code for scanning with smartphones would be cool. Maybe combined with some kind of remote-debugger.

**Reply by antfu** [reply]
Sounds like a good idea and should be straightforward to implement. Are you interested in make a PR? Thanks!

**Comment by Hebilicious** [comment]
Something that comes to mind would be to officially merge the spinoff vite project with this one https://github.com/webfansplz/vite-plugin-vue-devtools 
It would be nice if you could work together instead of fragmenting the ecosystem

**Reply by antfu** [reply]
We are not fragmenting them ecosystem, instead, we are exploring different possibilities in this realm. For sure we work together and will exchanges features we found are useful.

**Reply by Hebilicious** [reply]
That is nice to hear ! I am not familiar with both code bases, but would it be possible to have a monorepo integrating both the Nuxt and the Vite version? 
Don't get me wrong, this project that you created is incredible. I just don't want the vite version to take away the momentum from the Nuxt version, if that makes sense.

**Comment by marcoscode404** [comment]
it would be awesome if there was a way to parse API calls through the tool!

**Reply by antfu** [reply]
Can you elaborate a bit? Thanks

**Comment by Flowko** [comment]
> **Note**: This feature is supported in the latest version, with #291 ✅

would be nice to have the ability to view the image when an api route is returning a buffer
![image](https://github.com/nuxt/devtools/assets/35883748/1e9fc6e1-d32d-46a1-b520-932dbb4e0651)

**Reply by antfu** [reply]
This is supported via https://github.com/nuxt/devtools/pull/291

**Comment by franklin-tina** [comment]
Would be nice to have the ability to re-arrange the nav items.
For example, I mostly use these navs
- **pages**
- **server routes**
- **storage**
Would love to have them at the top, especially when the window is not too big, and they're hidden
<img width="370" alt="Screenshot 2023-06-27 at 07 09 44" src="https://github.com/nuxt/devtools/assets/76698025/03ec9a62-8c10-4591-a681-684471a906cd">

**Reply by antfu** [reply]
Supported!

**Comment by Smef** [comment]
I think it would be very convenient to have a display of the useFetch or useAsyncData calls which were executed during page load, how long they took, and how many bytes of data were returned from each one. This would be a great asset when trying to diagnose slow page render times.

**Comment by therealokoro** [comment]
Having server routes in the devtools was a killer idea. Just awesome, I use it all the time.
Coming from other rest clients and tools like postman and co. 

I can say that it would be nice to have the following.

- Variables that can be set to hold data which can be used in requests(all within the server routes tab)
- History of previous requests during a particular session, which can be fired again on click.

Who's with me✌🏾🖐🏾😏

**Reply by arashsheyda** [reply]
- the feature to have persistence variables was implemented with #321 
- can you explain a little more about the history of previous requests and it use cases? thanks

**Reply by tobychidi** [reply]
From my experience the Query, Body and Headers in the Devtools are not persistent or shareable. Is this possible?

**Reply by arashsheyda** [reply]
@tobychidi you can do that form `default inputs`

<img width="1393" alt="Screenshot 2023-12-05 at 3 49 00 PM" src="https://github.com/nuxt/devtools/assets/38922203/c6be4f91-9cab-4b95-8e62-22022e7860f8">

**Comment by segevfiner** [comment]
Would have been cool if this just integrated as a Vue devtools plugin so by just using Nuxt it automatically adds itself to the vue devtools like Pinia or vue-router do. But the Vue devtools API surface might be a bit limited for what the Nuxt devtools want to show for now, so Vue devtools might need to be expanded quite significantly to accommodate all it wants to do though this isn't a bad thing. Of course, this means the devtools eschews using it's own UI in the current design of the Vue devtools plugin API.

**Reply by antfu** [reply]
Indeed. Initially, Nuxt DevTools is designed to work with Vue DevTools, so we intentionally not covering the features Vue DevTools already have. As for now, we have a rough discussion with the Vue team, that maintaining Vue DevTools may become a bit challenging and we are seeking for a way to collaborate Nuxt DevTools, the browser extension's Vue DevTools, and the Vite plugin's Vue DevTools to work together with a shared infra. Nothing is settled yet but we will keep the community posted.

**Comment by Sevochka** [comment]
Would be nice to have a localStorage&cookie manager tool. The tool must support adding, removing and changing keys and values with a simple JSON editor interface. Something like [that](https://chrome.google.com/webstore/detail/swoosh-cookie-and-local-s/giompennnhheakjcnobejbnjgbbkmdnd) but also with mobile support

**Comment by Jak2k** [comment]
In Chrome Devtools you can open a folder. Then you can edit the styles and they will be saved back to the sources. Of course, this doesn't work with nuxt/vue. A way to visually edit the styles with this devtool and save them back to source would be really great.

**Comment by stephenjason89** [comment]
Would be nice to see and be able to make changes on the fly to the components & pages state. 
Similar to vue devtools so we don't need 2 devtools installed

<img width="1527" alt="image" src="https://github.com/nuxt/devtools/assets/3810177/825c33b7-870a-4e42-b274-7f6a7f36d04e">

**Comment by stephenjason89** [comment]
An ability to dock the devtool on the bottom, top, or sides would be nice as well.
This is to be able to see both your application and devtools without the devtools floating in the bottom center of your screen and covering content behind it.

**Comment by BayBreezy** [comment]
I think adding the ability to generate CRUD boilerplate code from the Server Routes tab in devtools would be great. Currently, we can view the api routes. Allowing users to enter the resource name and then have devtools create these 5 endpoints automagically would be nice:
<img width="297" alt="image" src="https://github.com/nuxt/devtools/assets/53905713/1b36e344-7d20-4fde-ad39-560df808b040">

This functionality is already at play in the Nuxt Mongoose module for nuxt: https://github.com/arashsheyda/nuxt-mongoose

**Reply by arashsheyda** [reply]
@BayBreezy Thank you for the suggestion! After discussing this with @antfu, we found his perspective to be quite insightful:

> 4\. I feel that might go about to far from Nuxt's scope. As even for CRUD, different project/company may have different conventions. I think it's better to leave it as 3rd party modules.

https://github.com/nuxt/devtools/issues/322#issuecomment-1643818818

I would be happy to assist you to create a custom module only for createing CRUDs if you want to.

**Reply by BayBreezy** [reply]
That would be great! The sad thing is, I don't know ow where to start in getting a client app to work with the devtools app lolol. I will do some digging and try to learn.i tried following the code in the nuxt-mongoose module client folder, but it seem complicated.

**Reply by arashsheyda** [reply]
that's great! you can DM me on discord if something came up (arashsheyda)

**Reply by casualmatt** [reply]
@antfu 
Relative to CRUD endpoints, and to #322 a good way to provide a flexible but easy way for module developers to add CRUD routes or any API endpoints, could be to have a new utility in the `@nuxt/devtools-kit` to add custom endpoints.

**Comment by mathieumagalhaes** [comment]
When adding a module, the list is currently ordered alphabetically.
It would be nice to be able to sort by:
- Popularity (npm installs)
- Newest (creation date)
- Stargazers (GitHub stars)
- Maintained (last modified date)

I wouldn't want real-time data, as that would require a lot of requests. 
Keep outgoing requests to a minimum.
An option would be the approach of Browserlist, to give user manual control of updates.

**Reply by arashsheyda** [reply]
implemented in #470. thank you!

**Reply by franklin-tina** [reply]
🔥🔥🔥

**Comment by MertcanYuzer** [comment]
Server Routes should allow an active/passive option for each choice in Query, Body, Headers, and Cookies tabs, so we don't have to delete and re-add them in each attempt.
I believe this would be useful when doing manual tests.
This feature is available in Postman, and I'm sharing sample screenshots below.
Active:
![image](https://github.com/nuxt/devtools/assets/47857239/60623133-c28c-43bd-a07f-50f5cea115bd)
Passive:
![image](https://github.com/nuxt/devtools/assets/47857239/9650da41-3733-41ba-b924-2e61ce5ef4bd)

**Reply by MertcanYuzer** [reply]
I saw that it was implemented in #474 . Thank you @arashsheyda

**Comment by khawarizmus** [comment]
**Inspect Provide and Inject**

I want to be able to see Where my provides are. Inspect them (probably modify them as well) and see where are they being injected and if they are.

https://vuejs.org/guide/components/provide-inject.html

**Comment by kucherenko** [comment]
**Server Logs, Events, Requests etc.**

It would be great to have the ability to track the server parameters like generated logs, executed SQL queries, HTTP requests, cloud function execution statistics, etc.

I think it would be helpful if the server could send special data to the client and the data will be displayed in the devtools.

Thank you

**Reply by Jak2k** [reply]
Maybe this would be better for a browser extension that integrates with vercel.

**Reply by kucherenko** [reply]
It can be helpful for different tech stacks, I think it can be part of the devtools as well

**Comment by nathanchase** [comment]
I'd love to be able to "pop out" Nuxt DevTools, similar to how you can undock the Inspect browser DevTools and have it in its own window.

<img width="342" alt="image" src="https://github.com/nuxt/devtools/assets/5963574/911df79b-62a7-4948-8d3c-9b1a2e9a2c2d">

I see "Popup", but clicking it doesn't appear to do anything currently?

<img width="351" alt="image" src="https://github.com/nuxt/devtools/assets/5963574/c92a8d55-ed78-4b0b-8af1-70b97c8bf79f">

**Reply by Jak2k** [reply]
Maybe your browser blocks it?

**Reply by nathanchase** [reply]
I have no other extensions running, other than Vue DevTools. Using Chrome 121.0.6113.0 .

**Reply by Jak2k** [reply]
Chrome and Firefox block pop-ups by default. There should be a small icon in the combo box (address bar) or you can change it inside the site permissions.

_P. S.: Consider using an [adblocker](https://ublockorigin.com/)._

**Reply by arashsheyda** [reply]
let's track it on #510

**Reply by nathanchase** [reply]
> _P. S.: Consider using an [adblocker](https://ublockorigin.com/)._

I do on my main browser. I use a separate browser install (https://www.google.com/chrome/canary/) specifically for dev work with zero extraneous extensions running.

**Reply by nathanchase** [reply]
Things to note about PiP that will likely not allow Nuxt DevTools to functionally work like an undocked browser DevTools:

> The Picture-in-Picture window in the Document Picture-in-Picture API is similar to a blank same-origin window opened via window.open(), with some differences:
>
>The Picture-in-Picture window floats on top of other windows.
>
>The Picture-in-Picture window never outlives the opening window.
>
>The Picture-in-Picture window cannot be navigated.
>
>The Picture-in-Picture window position cannot be set by the website.

https://developer.chrome.com/docs/web-platform/document-picture-in-picture/

Could we perhaps have Nuxt DevTools be a standalone Electron/Tauri app that can somehow connect to a browser instance - maybe through sockets? Then you would have full control of the Nuxt DevTools window.

**Reply by Smef** [reply]
I think a browser extension might be an easier thing to manage than a separate app.

**Comment by mathieumagalhaes** [comment]
I'd like to track performance of rendering/loading more effectively.

Currently we have, load/rendering took X seconds, and server-side rendering (SSR) took X seconds as well. 
It would be beneficial to have a more detailed analysis of this.

I'm thinking of the following:

- Total load/rendering timing (as it is currently displayed)
   - First Interactive timing 
      - Page First Byte timing
      - Timing of AsyncData calls (seperated by call/page/component)
      - Page Render Time
         - Time for rendering each component 
         - Time for hydrating each component
      - User Time
         - Time for each image to load
         
I'm probably missing other performance metrics.
In essence, this would be a simplified version of a flame graph using easy to understand Nuxt/Web terminology.

**Reply by ploca14** [reply]
I would love to be able to inspect the network requests made on the server during SSR. Similarly how you can in the browser devtools network tab. Making it easier to debug suspense waterfalls.

**Comment by Jak2k** [comment]
A view for linter output would be nice too.

**Comment by drobertson-dev** [comment]
Does anyone know if there is a way to track the state over time?
Currently I can see which functions are executed when via Timeline, but I can't see any way to view the change in values over time.  
I can see the initial payload but again I don't see any way to track the changes of those values.

If this isn't a feature already this is would be super helpful to have as a feature.

**Comment by martinszeltins** [comment]
My suggestion would be to make the Payload data more readable instead of randomly generated strings, it would be more useful to show the http endpoint.

![image](https://github.com/nuxt/devtools/assets/34019878/f6601e44-d063-401b-a42e-a3b86eda1862)

The other thing is that the data looks unreadable on dark mode:
![image](https://github.com/nuxt/devtools/assets/34019878/70dced5b-0ee1-4bc0-bc34-9671a3ee5f52)

**Comment by BayBreezy** [comment]
I think it would be nice to see a [vee-validate](https://vee-validate.logaretm.com/v4/) integration.

I saw where Pinia was added recently. Having the stuff from vee-validate in DevTools would be nice too.

**Reply by antfu** [reply]
3rd parties integrations should provided by the integrations themselves. You might want to make a request on their side.

**Reply by BayBreezy** [reply]
Ok I will.
@antfu Is there any resource out there that demonstrates how I can create my own little application and stick it into the devtools? I see a few modules do it here and there but I tried and I can't replicate it. I really want to create a video on how to do it but I can't find any resource that walks me through doing it from point A to Z. 
It would be nice to have a guide on how to do that. That way maybe I can create the vee-validate integration

**Reply by arashsheyda** [reply]
@BayBreezy https://devtools.nuxt.com/module/guide, also feel free to message me on Discord or Twitter

**Comment by KevinvOosterhout** [comment]
Next to the integration with VSCode i would love an integration with Webstorm/Jetbrains IDEs. Me (and a lot of other people) are exclusively working inside of the Jetbrains IDEs

**Comment by dsvgl** [comment]
Could we have an option for the devTools modal to have a drop shadow?
I use it in light mode and my website is mostly white. It looks kinda being part of the website and not laying on top of it 🙃

**Reply by arashsheyda** [reply]
for now you can modify the style using `#nuxt-devtools-iframe`
e.g:
```css
#nuxt-devtools-iframe {
  box-shadow: 0 0 10px rgba(0px, 0px, 10px, 0.3);
}
```

**Reply by dsvgl** [reply]
Thx! This works.
I added a custom css file to the `css` array inside the `$development` property of nuxt.config.ts. 🤌🏻

**Comment by jiakun-zhao** [comment]
DevTools 有在新的浏览器标签页打开的方式吗？（不在我的页面显示按钮打开，类似 UnoCSS 在 Vite 上那样）

**Comment by blqke** [comment]
Are we able to configure UI settings locally ? 


Because every time I start a new Nuxt project, I find myself tweaking the same DevTools settings (check "Close DevTools when clicking outside", check "Expand Sidebar", check "Scrollable sidebar", etc).


Maybe I missed something, but is there a way to configure default ui settings this way in `nuxt.config.ts` ?
```
 devtools: {
    enabled: true,
    ui: {
      interactionCloseOnOutsideClick: true,
    },
 }
```

or maybe this way, locally via `.nuxtrc` ?
```
devtools.ui.interactionCloseOnOutsideClick=true
```

**Reply by ManUtopiK** [reply]
UI settings are stored in the browser. So if you start a new project on `localhost:3000` you should retrieve your config.

**Reply by martinszeltins** [reply]
Storing it in the browser is not a good idea, every time you clear your browser you lose the settings.

**Comment by kingtimm** [comment]
Seeing [storybooks component mounting and testing](https://storybook.js.org/blog/storybook-8-3/) using vitest makes me wonder about the ability to have the devtools components panel enhanced to allow mounting of a component with editable props. Pages devtools panel could even use route params/query in addition to props.

**Comment by ThomasBerneHCSE** [comment]
Our team is developing inside a docker container and code doesn't exist in that container but on the machine hosting it. it means that the inspector is trying to open code inside the container because the `npm run dev` is inside it. Is it possible to add an option to copy the component path to the clipboard on click of the inspect instead of trying to open code with it ?

**Reply by martinszeltins** [reply]
It would be even better to teach it to open the file in vscode automatically by providing the path

**Comment by tojoshan** [comment]
Show env variables.

**Reply by martinszeltins** [reply]
You can already see runtime config which is overriden with env variables.

**Comment by GPlemon** [comment]
"The 'click to go to file' functionality should be modified to:

First check if VS Code is already running inside dev-tools via code-server
If VS Code is running in dev-tools, open the file there instead of locally
If VS Code is not running in dev-tools, open the file locally without launching dev-tools (which is happening currently)"

This change would create a more intuitive workflow where files always open in the most appropriate editor context without unnecessary dev-tools launches.

**Comment by ghost** [comment]
I think it would be nice to add some features from Nuxt Studio for CMS editing, like components replacing or smth like that. I know that idea not so good enough, but i think for some developers will be interesting to manipulate components from page and update their. Also, i think would be nice to add i18n supporting for analyzing and manipulating that

**Comment by arbitino** [comment]
devtools for useState, with reactive updates and call history, if possible

**Comment by daryllxd** [comment]
Is there a way for the DevTools to always be in the pop-up mode? (ideally the nuxt tools would be in another window, esp for multiple monitors. The Because what happens is that every time I refresh my browser, it's still stuck in the non pop-up mode and I need to click it. There's not even a keyboard shortcut to pop up.

![CleanShot 2025-05-23 at 12 31 05](https://github.com/user-attachments/assets/13042380-a21e-4791-af14-589df3f8a29b)


I tried adding a client-side plugin but no go. I suspect it's in https://github.com/nuxt/devtools/blob/fedaa13501b23357e171b56827585f773e9c5202/packages/devtools/client/setup/vue-devtools.ts#L8

**Reply by martinszeltins** [reply]
There is option to open in Popup window

**Reply by daryllxd** [reply]
I'm so sorry man if I am blind (trust me I am spamming cmd-F "popup" here in the **Settings tab - is it somewhere else? I've checked every single tab here**)  but here is what it looks like when I navigate through the settings page after I've opened the dev tools. 

https://github.com/user-attachments/assets/ce40040e-300d-4642-82f7-41d5008c5411

**Reply by martinszeltins** [reply]
yea doesn't look like you can configure it to always open in popup. But you can create an issue / feature request, that would be a cool feature to have for sure

**Comment by Pentadome** [comment]
debugging while rendering with ssr is way harder. Would love to see a quick toggle between rendering with ssr or without at the bottom bar (before you open the modal)

**Comment by saschafuchs** [comment]
I would also like to use Nuxt Devtools for builds, e.g. within the Dev / Stage Server. Because when we integrate SPAs, it would be nice to see what the data looks like in the stores and in the payload, what is connected to the components. This was already available in Nuxt 2 with the old vue devtools.

**Comment by Desarrollo-8** [comment]
Add the option to view Pinia directly to the dev tools

**Comment by jd-solanki** [comment]
How about having AI Integrations for cases like nuxt/hints hydration mismatch can help resolve the issues via AI? 🤔 

https://github.com/nuxt/hints/issues/124

> _P.S. Just thinking out loud_

**Comment by segevfiner** [comment]
I wonder why the timeline from the official Vue devtools is missing...
